### PR TITLE
The initial byte doesn't count to the remaining length

### PR DESCRIPTION
**Pull request by rasdie**
The Initial Byte of the MQTT message does not count to the Remaining Length. As result the calculation of the message length is wrong if the Remaining Length is exactly 127 bytes (or 16383, ...). As result the MQTT broker (in my case mosquitto) terminates the connection to the MQTT client (Client ESP8266-XXXXXX disconnected due to protocol error.).

### CODE REVIEW

File: src/serialiser.c [hunk 1]
```diff
@@ -191,6 +190,8 @@ size_t mqtt_serialiser_size(mqtt_serialiser_t* serialiser, mqtt_message_t* messa
     len += 4;
   }
 
+  len += 1; // initial byte with flags and type
```
Comment (slaff) [inline src/serialiser.c:193]: Can you revert the change in this line? The rest looks good.

**Comment by slaff** [comment]
@rasdie Please add a test case for the fix in https://github.com/slaff/mqtt-codec/blob/master/bin/test.c. We will use the test case to prevent regressions.

File: src/serialiser.c (hunk 1 above)
Comment (rasdie) [reply src/serialiser.c:193]: Without that line all MQTT telegrams are to short by one byte. len counts the bytes of the complete telegram.

Comment (slaff) [reply src/serialiser.c:193]: Thanks for the tests. Now concerning this simple change - may be I am missing something but I don't see a  difference in the logic between the old code, apart from moving the `len+=1;` line after the if-block:

```C
  message->common.length = len;
  len += 1; // initial byte with flags and type

  // the bytes needed for the common length
  if(len <= 127) {
    len += 1;
  } else if(len <= 16383) {
    len += 2;
  } else if(len <= 2097151) {
    len += 3;
  } else if(len <= 268435455) {
    len += 4;
  }

  return len;
```

and the new one:
```C
  message->common.length = len;

  // the bytes needed for the common length
  if(len <= 127) {
    len += 1;
  } else if(len <= 16383) {
    len += 2;
  } else if(len <= 2097151) {
    len += 3;
  } else if(len <= 268435455) {
    len += 4;
  }

  len += 1; // initial byte with flags and type

return len;
```

If there is such a difference please explain it to me, if not just revert the change(return the len+1 line to its old position).

File: bin/test.c [hunk 2]
```diff
@@ -190,52 +190,101 @@ void test_parser() {
   printf("\n");
 }
 
+static void test_serialiser_procces_msg(mqtt_serialiser_t* serialiser, mqtt_message_t* message) {
```
Comment (slaff) [inline bin/test.c]: Rename the function to `test_serialiser_proccess_message`.

File: src/serialiser.c (hunk 1 above)
Comment (rasdie) [reply src/serialiser.c:193]: Hi slaff,
sure, for the most cases both versions are identical! But look for the special cases 127, 2097151, and 2097151:
In the old version if len is 127 at the beginning (= the remaining length or common length), then len will be increased  by one **before** calculating the size of the remaining length field. This results in a size of **2 bytes** and in an overall packet length of 130 bytes.
In my version the size of the remaining length field is only **1 byte** (correct to decode a length of 127 bytes). The packet length becomes 129 bytes.
Best regards,
Rasmus

Comment (slaff) [reply src/serialiser.c:193]: Ah.. I see now (said the blind man). Thanks for the clarification.

File: bin/test.c (hunk 2 above)
Comment (slaff) [reply bin/test.c]: > Rename the function to test_serialiser_proccess_message.

@rasdie You have marked this as resolved but I don't see the change? Can you please commit it?

**Review by rasdie — COMMENTED** [review]
Review comment adapted
